tests/php/Emitter/OutputEmitter/NodeEmitter/ApplyEmitterTest.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PhelTest\Emitter\OutputEmitter\NodeEmitter;
+
+use Phel\Ast\ApplyNode;
+use Phel\Ast\DefNode;
+use Phel\Ast\LiteralNode;
+use Phel\Ast\PhpVarNode;
+use Phel\Ast\TupleNode;
+use Phel\Emitter\OutputEmitter;
+use Phel\Emitter\OutputEmitter\Munge;
+use Phel\Emitter\OutputEmitter\NodeEmitter\ApplyEmitter;
+use Phel\Emitter\OutputEmitter\NodeEmitterFactory;
+use Phel\Emitter\OutputEmitter\SourceMap\SourceMapGenerator;
+use Phel\Lang\Symbol;
+use Phel\Lang\Table;
+use Phel\NodeEnvironment;
+use PHPUnit\Framework\TestCase;
+
+final class ApplyEmitterTest extends TestCase
+{
+    private ApplyEmitter $applyEmitter;
+
+    public function setUp(): void
+    {
+        $this->applyEmitter = new ApplyEmitter(
+            new OutputEmitter(
+                $enableSourceMaps = true,
+                new SourceMapGenerator(),
+                new NodeEmitterFactory(),
+                new Munge()
+            )
+        );
+    }
+
+    public function testPhpVarNodeAndFnNodeIsInfix(): void
+    {
+        $node = new PhpVarNode(NodeEnvironment::empty(), '+');
+        $args = [
+            new TupleNode(NodeEnvironment::empty(), [
+                new LiteralNode(NodeEnvironment::empty(), 2),
+                new LiteralNode(NodeEnvironment::empty(), 3),
+                new LiteralNode(NodeEnvironment::empty(), 4),
+            ]),
+        ];
+
+        $arrayNode = new ApplyNode(NodeEnvironment::empty(), $node, $args);
+        $this->applyEmitter->emit($arrayNode);
+
+        $this->expectOutputString(
+            'array_reduce([...((\Phel\Lang\Tuple::createBracket(, , );) ?? [])], function($a, $b) { return ($a + $b); });'
+        );
+    }
+
+    public function testPhpVarNodeButNoInfix(): void
+    {
+        $node = new PhpVarNode(NodeEnvironment::empty(), 'str');
+        $args = [
+            new LiteralNode(NodeEnvironment::empty(), 'abc'),
+            new LiteralNode(NodeEnvironment::empty(), 'def'),
+        ];
+
+        $arrayNode = new ApplyNode(NodeEnvironment::empty(), $node, $args);
+        $this->applyEmitter->emit($arrayNode);
+
+        $this->expectOutputString('str(, ...(() ?? []));');
+    }
+
+    public function testNoPhpVarNode(): void
```

Comment:
No this is not what you wanted, sorry ;)

This is something like
```
(apply (defn foo [arg] (println arg)) [true])
```

Which branch/line do you want to cover with this test. Maybe I can give you a better example.

Reply:
I want to test the `$this->noPhpVarNode($node)` method (last one).

In other words, I need to test the branch when the `$node->getFn()` won't be an instance of `PhpVarNode` class.

That's why I tried but not sure if the example fits either how to write it. :stuck_out_tongue: 
```clojure
>>> (defn foo [arg] (println arg))
1
>>> (apply foo [true])
true
nil
>>> 
```
Any suggestion?

Reply:
I would recommend this:

```
(apply (fn [& x] x) [1])
```

but this one is even easier (but doesn't make much sense)

```
(apply (fn [&]) [1])
```

With can be expressed as:

```
$tuple = Tuple::create(
  Symbol::create('apply'),
  Tuple::create(
    Symbol::create('fn'),
    Tuple::createBracket(
      Symbol::create('&')
    ),
  ),
  Tuple::createBracket(
    1
  )
)
```

Reply:
@jenshaase I have some troubles.

`ApplyEmitter` only `emit()` `Node` objects, so, I guess it should emit the `ApplyNode` object.
The `ApplyNode` constructor needs `NodeEnvironment`, `Node $fn`, `array $arguments` and an optional `SourceLocation`.
I can pass a `NodeEnvironment::empty()` as the first argument, but I don't know what to do with the `Node $fn`.
I tried with a `FnNode` but it needs a lot of parameters (one of them is also another node...). I am not sure if this is right.

In your example, you are creating a Tuple instead of a Node and I don't know how to implement it into the test. Could you take a look, please? :smiley:

Reply:
Yeah, sorry I was in the wrong context. But you figured it out yourself.